lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,287 @@
+    #         3. group(optional):  The name of the External IPAM group containing the subnet.
+    #
+    # Returns: true if IP was added successfully to External IPAM, otherwise false
+    #
+    # Responses from Proxy plugin:
+    #   Response if success:
+    #     {"code": 200}
+    #   Response if IP already reserved:
+    #     {"code": 409, "error": "IP address already exists"}
+    #   Response if subnet error:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if missing required params:
+    #     {"code": 400, "error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","Missing 'ip' parameter. An IPv4 address must be provided(e.g. 100.10.10.22)"]}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def add_ip_to_subnet(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse post({}, "/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise(response['error']) if response['error'].present?
+      response['code'] == 200
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to add IP %{ip} to the subnet %{subnet} in External IPAM."), ip: ip, subnet: subnet)
+    end
+
+    # Get a list of groups from External IPAM. A group is analagous to a 'section' in phpIPAM, and
+    # is a logical grouping of subnets/ips.
+    #
+    # Params: None
+    #
+    # Returns: An array of groups on success, or a hash with a "message" key
+    #          containing error on failure.
+    #
+    # Responses from Proxy plugin:
+    #   Response if success:
+    #     {"code": 200, "data": [
+    #       {name":"Test Group","description": "A Test Group"},
+    #       {name":"Awesome Group","description": "A totally awesome Group"}
+    #     ]}
+    #   Response if no groups exist:
+    #     {"code": 200, "data": []}
+    #   Response if groups are not supported:
+    #     {"code": 501, "error": "Groups are not supported"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def get_groups
+      response = parse get("/groups")
+      raise(response['error']) if response['error'].present?
+      response
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain groups from External IPAM."))
+    end
+
+    # Get a group from External IPAM. A group is analagous to a 'section' in phpIPAM, and
+    # is a logical grouping of subnets/ips.
+    #
+    # Params: group: The name of the External IPAM group
+    #
+    # Returns: An External IPAM group on success, or a hash with an "error" key on failure.
+    #
+    # Responses from Proxy plugin:
+    #   Response if success:
+    #     {"code": 200, "data": {"name":"Awesome Section", "description": "Awesome Section"}}
```

Comment:
I wonder: is this the literal response JSON or a meta-description? I assumed the latter that the body is actually:
```json
{"name":"Awesome Section", "description": "Awesome Section"}
```

And HTTP status code is 200. Now that I read `subnet_exists_in_group` again, it uses `group['data'].present?` and that makes me doubt it. In a REST API I wouldn't expect a wrapper with the status code since at the HTTP level those codes are present.

Reply:
Yeah I actually missed that too, can you refactor this one more time? This one is pretty straightforward - just extract "code" from body to be HTTP request code both in core and in the plugin.

Normally, I'd merge this and ask you to do in a separate PR, but here we actually define how the API should look like and there was already one community user interested in writing own implementation. Let's get this right from the day one. Tests are green btw the one that fails is a random failure we currently face.